### PR TITLE
Read-only sandbox paths mounts (Linux)

### PR DESCRIPTION
**Pull request by SimSaladin**
## Motivation

This extends the `sandbox-paths` syntax to allow the extra optional suffix `:ro` for marking that path so as to be mounted read-only in the sandbox. The suffix is excluded from the mounted source or target path. It will only cause the mount `MS_RDONLY` flag to be set.

## Context

...well, technically we make then another `mount(2)` call on the target path right after mounting the path for the first time, in order to do a second mount (or rather remount, `MS_REMOUNT`) with `MS_RDONLY`. For some reason single `mount` syscall can't both create a new bind-mount and assign the read-only flag to it (its mount-point really) ... or propagation properties for that matter.

The `mount(8)` cli from `util-linux` abstracts over this with a second syscall as necessary as well (more even sometimes, it seems).

There's a better, modern API since about linux 5.12 with `open_tree()`, `mount_setattr()` and `move_mount()`. Probably doesn't really make a difference until there's some need or want for those features though.

---

One of the commits (2eaca522e8adecfcbb3dd96c853ede26c5e484cc) is not related to to the bulk of the changes in this. it's a small change to fix a certain edge-case (think user launching `nix` or running tests with a `PATH` that still works, for all intents and purposes, but looks like it fell through a slipstream portal sideways).

---

add :+1: to [pull requests you find important](https://github.com/nixos/nix/pulls?q=is%3aopen+sort%3areactions-%2b1-desc).

the nix maintainer team uses a [github project board](https://github.com/orgs/nixos/projects/19) to [schedule and track reviews](https://github.com/nixos/nix/tree/master/maintainers#project-board-protocol)

### CODE REVIEW

File: src/libstore/unix/build/linux-derivation-builder.cc [hunk 1]
```diff
@@ -121,13 +121,18 @@ static void setupSeccomp()
...
     debug("bind mounting '%1%' to '%2%'", source, target);
 
     auto bindMount = [&]() {
         if (mount(source.c_str(), target.c_str(), "", MS_BIND | MS_REC, 0) == -1)
             throw SysError("bind mount from '%1%' to '%2%' failed", source, target);
+
+        if (rdonly)
+            // initial mount wouldn't respect MS_RDONLY, must remount
+            if (mount("", target.c_str(), "", MS_REMOUNT | MS_BIND | MS_RDONLY, 0) == -1)
+                throw (SysError("making bind mount '%s' read-only failed", target));
```
Comment (edolstra) [inline src/libstore/unix/build/linux-derivation-builder.cc]:
```suggestion
                throw SysError("making bind mount '%s' read-only failed", target);
```

File: src/libstore/unix/build/linux-derivation-builder.cc [hunk 2]
```diff
@@ -121,13 +121,18 @@ static void setupSeccomp()
 #  endif
 }
 
-static void doBind(const Path & source, const Path & target, bool optional = false)
+static void doBind(const Path & source, const Path & target, bool optional = false, bool rdonly = false)
```
Comment (edolstra) [inline src/libstore/unix/build/linux-derivation-builder.cc]:
```suggestion
static void doBind(const Path & source, const Path & target, bool optional = false, bool readOnly = false)
```

File: src/libstore/include/nix/store/globals.hh [hunk 3]
```diff
@@ -632,13 +632,18 @@ public:
...
         R"(
-          A list of paths bind-mounted into Nix sandbox environments. You can
-          use the syntax `target=source` to mount a path in a different
-          location in the sandbox; for instance, `/bin=/nix-bin` will mount
-          the path `/nix-bin` as `/bin` inside the sandbox. If *source* is
-          followed by `?`, then it is not an error if *source* does not exist;
-          for example, `/dev/nvidiactl?` specifies that `/dev/nvidiactl` will
-          only be mounted in the sandbox if it exists in the host filesystem.
+          A list of paths bind-mounted into Nix sandbox environments. Use the
+          syntax `target[=source][:ro][?]` to control the mount:
```
Comment (edolstra) [inline src/libstore/include/nix/store/globals.hh]: Instead of ad hoc formats like this, I would prefer that structured options use JSON, e.g.
```
sandbox-paths = [ {"target": "/foo", "source": "/bar", "readOnly": true, "optional": true} ]
```
or perhaps
```
sandbox-paths = { "/foo" : {"source": "/bar", "readOnly": true, "optional": true} }
```

This also makes it more future-proof when we switch `nix.conf` to JSON entirely.

For an example of how to do JSON options, see the `external-builders` setting in this PR:
* https://github.com/DeterminateSystems/nix-src/pull/78/files#diff-fc61af5dc32777fbdca32b2c72c3a1842457dc92dfcf47919bd96b944b7a6479
* https://github.com/DeterminateSystems/nix-src/pull/78/files#diff-4f84c7126ef8b87cefd2d430990cceb5c920e1684f1e9c28bbb4104bd7632304

In this case it's slightly more complex because we have to handle the non-JSON case for backward compatibility, but that should be pretty easy (just check if the first non-whitespace character is `[`).

Comment (SimSaladin) [reply src/libstore/include/nix/store/globals.hh]: Implemented this and the smaller changes in latest commit, please take a look.

File: src/libstore/include/nix/store/globals.hh [hunk 4]
```diff
@@ -18,6 +21,72 @@ namespace nix {
...
+        slave = MS_SLAVE,
+        unbindable = MS_UNBINDABLE
+#else
+        ro  // FIXME: do any options make sense on other that linux?
+#endif
+    } MountOpt;
+
+#ifdef __linux__
+    /* Options to set when readOnly=true */
+    constexpr static MountOpt readOnlyDefaults[] = { ro, nodev, noexec, nosuid, noatime };
```
Comment (edolstra) [inline src/libstore/include/nix/store/globals.hh]: Hm, seems odd to have "readOnly" imply `noexec`, `noatime` etc. I would expect a readonly mount to still be executable by default.

File: src/libstore/include/nix/store/globals.hh [hunk 5]
```diff
@@ -18,6 +21,72 @@ namespace nix {
...
+        nodev = MS_NODEV,
+        noexec = MS_NOEXEC,
+        nosuid = MS_NOSUID,
+        noatime = MS_NOATIME,
+        nodiratime = MS_NODIRATIME,
+        relatime = MS_RELATIME,
+        strictatime = MS_STRICTATIME, /* overrides any atime/relatime */
+        private_ = MS_PRIVATE,
+        slave = MS_SLAVE,
+        unbindable = MS_UNBINDABLE
```
Comment (edolstra) [inline src/libstore/include/nix/store/globals.hh]: Not sure if all of these mount options make sense to expose. In particular, the mount propagation flags seem a bit scary. We don't want the builder to be able to have mounts propagate out of the sandbox...

We should probably set `MS_NOSUID` unconditionally since it's nice for security. Though unfortunately it doesn't prevent the creation of suid binaries.

Comment (SimSaladin) [reply src/libstore/include/nix/store/globals.hh]: Propagation outside the sandbox isn't possible regardless of the propagation type AFAICT... There is no propagation tree shared between the sandbox and host once `/` is mounted `MS_PRIVATE | MS_REC` in a mount namespace, so the effects of any propagation flag is limited to the sandbox. Admittedly probably not a very useful setting agreed.

Perhaps being able to do what docker has as e.g. `--mount type=bind,bind-propagation=rslave` could be a little bit less useless, though?

`nosuid` by default sounds good.

File: src/libstore/include/nix/store/globals.hh (hunk 4 above)
Comment (SimSaladin) [reply src/libstore/include/nix/store/globals.hh]: I initially followed the approach used by `runc` when handling the `ReadOnlyPaths` option in the OCI spec, which seemed reasonable. However, it turns out that this logic differs from what's applied to read-only volumes and the `docker run --read-only` flag. Given that, I’ll revert the change and stick with using just `MS_RDONLY`.

File: .version [hunk 6]
```diff
@@ -1 +1 @@
-2.31.0
+2.30.0
```
Comment (Mic92) [inline .version]: Needs to be removed.
